pkg/sentry/fsimpl/iouringfs/iouringfs.go

```diff
@@ -0,0 +1,323 @@
+// discrepancy from Linux.
+func (iouringFSType) Name() string {
+	return "iouring"
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (iouringFSType) Release(ctx context.Context) {}
+
+// +stateify savable
+type iouringFS struct {
+	kernfs.Filesystem
+
+	devMinor uint32
+}
+
+// NewFilesystem sets up and returns a new iouring filesystem.
+//
+// Note that there should only ever be one instance of iouring.Filesystem,
+// backing a global io_uring mount.
+func NewFilesystem(vfsObj *vfs.VirtualFilesystem) (*vfs.Filesystem, error) {
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		return nil, err
+	}
+
+	fs := &iouringFS{
+		devMinor: devMinor,
+	}
+
+	fs.Filesystem.VFSFilesystem().Init(vfsObj, iouringFSType{}, fs)
+
+	return fs.Filesystem.VFSFilesystem(), nil
+}
+
+// Release implements vfs.FilesystemImpl.Release.
+func (fs *iouringFS) Release(ctx context.Context) {
+	fs.Filesystem.VFSFilesystem().VirtualFilesystem().PutAnonBlockDevMinor(fs.devMinor)
+	fs.Filesystem.Release(ctx)
+}
+
+// PrependPath implements vfs.FilesystemImpl.PrependPath.
+func (fs *iouringFS) PrependPath(ctx context.Context, vfsroot, vd vfs.VirtualDentry, b *fspath.Builder) error {
+	inode := vd.Dentry().Impl().(*kernfs.Dentry).Inode().(*inode)
+	b.PrependComponent(fmt.Sprintf("iouring:[%d]", inode.InodeAttrs.Ino()))
+	return vfs.PrependPathSyntheticError{}
+}
+
+// MountOptions implements vfs.FilesystemImpl.MountOptions.
+func (fs *iouringFS) MountOptions() string {
+	return ""
+}
+
+// inode implements kernfs.Inode.
+//
+// +stateify savable
+type inode struct {
+	kernfs.InodeAttrs
+	kernfs.InodeNoopRefCount
+	kernfs.InodeNotDirectory
+	kernfs.InodeNotSymlink
+}
+
+// Open implements kernfs.Inode.Open. iouringfs inode can not be opened
+func (i *inode) Open(ctx context.Context, rp *vfs.ResolvingPath, d *kernfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	return nil, syserror.EFAULT
+}
+
+// StatFS implements kernfs.Inode.StatFS.
+func (i *inode) StatFS(ctx context.Context, fs *vfs.Filesystem) (linux.Statfs, error) {
+	return vfs.GenericStatFS(linux.SOCKFS_MAGIC), syserror.EFAULT
+}
+
+// NewDentry constructs and returns a iouringfs dentry.
+//
+// Preconditions: mnt.Filesystem() must have been returned by NewFilesystem().
+func NewDentry(ctx context.Context, mnt *vfs.Mount) *vfs.Dentry {
+	fs := mnt.Filesystem().Impl().(*iouringFS)
+
+	filemode := linux.FileMode(linux.S_IFREG | 0600)
+	i := &inode{}
+	i.InodeAttrs.Init(ctx, auth.CredentialsFromContext(ctx), linux.UNNAMED_MAJOR, fs.devMinor, fs.Filesystem.NextIno(), filemode)
+
+	d := &kernfs.Dentry{}
+	d.Init(&fs.Filesystem, i)
+	return d.VFSDentry()
+}
+
+// fileDescription is embedded by iouringfs implementations of
+// vfs.FileDescriptionImpl.
+//
+// +stateify savable
+type fileDescription struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.LockFD // FIXME: This should probably be a LockFD
```

Comment:
I think so. Yo never know what userland programs may reserve. For example if 2 threads call ``io_uring_register`` this might modify the file description. Or if 2 threads are trying to write to the same memory page which is not currently mapped